internal/extproc/translator/openai_gcpanthropic.go

```diff
@@ -9,90 +9,648 @@
+			case contentPart.ImageContent != nil:
+				imageURL := contentPart.ImageContent.ImageURL.URL
+				if isDataURI(imageURL) {
+					contentType, data, err := parseDataURI(imageURL)
+					if err != nil {
+						return nil, fmt.Errorf("failed to parse image URL: %w", err)
+					}
+					appPDF := constant.ValueOf[constant.ApplicationPDF]()
+					base64Data := base64.StdEncoding.EncodeToString(data)
+					if contentType == string(appPDF) {
+						pdfSource := anthropic.Base64PDFSourceParam{
+							Data: base64Data,
+						}
+						resultContent = append(resultContent, anthropic.NewDocumentBlock(pdfSource))
+					} else if isSupportedImageMediaType(contentType) {
+						resultContent = append(resultContent, anthropic.NewImageBlockBase64(contentType, base64Data))
+					} else {
+						return nil, fmt.Errorf("invalid media_type for image '%s'", contentType)
+					}
+				} else if strings.HasSuffix(strings.ToLower(imageURL), ".pdf") {
+					resultContent = append(resultContent, anthropic.NewDocumentBlock(anthropic.URLPDFSourceParam{
+						URL: imageURL,
+					}))
+				} else {
+					resultContent = append(resultContent, anthropic.NewImageBlock(anthropic.URLImageSourceParam{
+						URL: imageURL,
+					}))
+				}
+			case contentPart.InputAudioContent != nil:
+				return nil, fmt.Errorf("input audio content not supported yet")
+			}
+		}
+		return resultContent, nil
+	case openai.StringOrArray:
+		switch val := v.Value.(type) {
+		case string:
+			if val == "" {
+				return nil, nil
+			}
+			return []anthropic.ContentBlockParamUnion{
+				anthropic.NewTextBlock(val),
+			}, nil
+		case []openai.ChatCompletionContentPartUserUnionParam:
+			return openAIToAnthropicContent(val)
+		default:
+			return nil, fmt.Errorf("unsupported StringOrArray value type: %T", val)
+		}
+	}
+	return nil, fmt.Errorf("unsupported OpenAI content type: %T", content)
+}
+
+func extractSystemOrDeveloperPromptFromSystem(msg openai.ChatCompletionSystemMessageParam) string {
+	switch v := msg.Content.Value.(type) {
+	case string:
+		return v
+	case []openai.ChatCompletionContentPartUserUnionParam:
+		// Concatenate all text parts for completeness
+		var sb strings.Builder
+		for _, part := range v {
+			if part.TextContent != nil {
+				sb.WriteString(part.TextContent.Text)
+			}
+		}
+		return sb.String()
+	case nil:
+		return ""
+	default:
+		// If msg.Content is a StringOrArray, unwrap it
+		if soarr, ok := msg.Content.Value.(openai.StringOrArray); ok {
+			switch val := soarr.Value.(type) {
+			case string:
+				return val
+			case []openai.ChatCompletionContentPartUserUnionParam:
+				var sb strings.Builder
+				for _, part := range val {
+					if part.TextContent != nil {
+						sb.WriteString(part.TextContent.Text)
+					}
+				}
+				return sb.String()
+			}
+		}
+	}
+	return ""
+}
+
+func extractSystemOrDeveloperPromptFromDeveloper(msg openai.ChatCompletionDeveloperMessageParam) string {
```

Comment:
can use `systemMsgToDeveloperMsg` to avoid duplicating extractSystemOrDeveloperPromptFromSystem and extractSystemOrDeveloperPromptFromDeveloper 


systemMsgToDeveloperMsg:  https://github.com/sukumargaonkar/ai-gateway/blob/18d9c13c007b50f1433ce75abbe8b208dfc96b1c/internal/extproc/translator/gemini_helper.go#L114-L121

Reply:
can you move this to general util file?

Reply:
updated